### ISSUE TITLE
DbContext.X.Where(x => x.YRef.ID == 1) throws "The binary operator Equal is not defined for the types Nullable..."

### ISSUE THREAD

**Issue by LonDat** [aspnet/EntityFramework#4036]
Classes:
```C#
public class X {
  public int ID { get; set; }
  public Y YRef { get; set; }
}
public class Y {
  public int ID { get; set; }
}
```
Using (exception when):
```
X xxx = dbContext.Xs.Single(x => x.YRef.ID == 1);
```
Exception:
```
System.InvalidOperationException occurred
  HResult=-2146233079
  Message=The binary operator Equal is not defined for the types 'System.Nullable`1[System.Int32]' and 'System.Int32'.
  Source=System.Core
  StackTrace:
       at System.Linq.Expressions.Expression.GetEqualityComparisonOperator(ExpressionType binaryType, String opName, Expression left, Expression right, Boolean liftToNull)
       at System.Linq.Expressions.Expression.Equal(Expression left, Expression right, Boolean liftToNull, MethodInfo method)
       at System.Linq.Expressions.Expression.MakeBinary(ExpressionType binaryType, Expression left, Expression right, Boolean liftToNull, MethodInfo method, LambdaExpression conversion)
       at System.Linq.Expressions.Expression.MakeBinary(ExpressionType binaryType, Expression left, Expression right, Boolean liftToNull, MethodInfo method)
       at Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitBinary(BinaryExpression node) in C:\LonDatSources\Git\Repos\EntityFramework\src\EntityFramework.Core\Query\ExpressionVisitors\Internal\NavigationRewritingExpressionVisitor.cs:line 225
  InnerException: 

```
Callstack from debugging:
```
 	System.Core.dll!System.Linq.Expressions.Expression.GetEqualityComparisonOperator(System.Linq.Expressions.ExpressionType binaryType, string opName, System.Linq.Expressions.Expression left, System.Linq.Expressions.Expression right, bool liftToNull)	Unknown
 	System.Core.dll!System.Linq.Expressions.Expression.Equal(System.Linq.Expressions.Expression left, System.Linq.Expressions.Expression right, bool liftToNull, System.Reflection.MethodInfo method)	Unknown
 	System.Core.dll!System.Linq.Expressions.Expression.MakeBinary(System.Linq.Expressions.ExpressionType binaryType, System.Linq.Expressions.Expression left, System.Linq.Expressions.Expression right, bool liftToNull, System.Reflection.MethodInfo method, System.Linq.Expressions.LambdaExpression conversion)	Unknown
 	System.Core.dll!System.Linq.Expressions.Expression.MakeBinary(System.Linq.Expressions.ExpressionType binaryType, System.Linq.Expressions.Expression left, System.Linq.Expressions.Expression right, bool liftToNull, System.Reflection.MethodInfo method)	Unknown
>	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.VisitBinary(System.Linq.Expressions.BinaryExpression node = {System.Linq.Expressions.LogicalBinaryExpression}) Line 225	C#
 	System.Core.dll!System.Linq.Expressions.BinaryExpression.Accept(System.Linq.Expressions.ExpressionVisitor visitor)	Unknown
 	System.Core.dll!System.Linq.Expressions.ExpressionVisitor.Visit(System.Linq.Expressions.Expression node)	Unknown
 	Remotion.Linq.dll!Remotion.Linq.Clauses.WhereClause.TransformExpressions(System.Func<System.Linq.Expressions.Expression, System.Linq.Expressions.Expression> transformation)	Unknown
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.NavigationRewritingQueryModelVisitor.VisitWhereClause(Remotion.Linq.Clauses.WhereClause whereClause = {Remotion.Linq.Clauses.WhereClause}, Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel}, int index = 0) Line 597	C#
 	Remotion.Linq.dll!Remotion.Linq.Clauses.WhereClause.Accept(Remotion.Linq.IQueryModelVisitor visitor, Remotion.Linq.QueryModel queryModel, int index)	Unknown
 	Remotion.Linq.dll!Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(System.Collections.ObjectModel.ObservableCollection<Remotion.Linq.Clauses.IBodyClause> bodyClauses, Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel})	Unknown
 	Remotion.Linq.dll!Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel})	Unknown
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.NavigationRewritingExpressionVisitor.Rewrite(Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel}) Line 102	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.EntityQueryModelVisitor.OptimizeQueryModel(Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel}) Line 231	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.EntityQueryModelVisitor.CreateQueryExecutor<LonDatSources.Mail.Service.Entity.EmailBoxAddress>(Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel}) Line 153	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Storage.Database.CompileQuery<LonDatSources.Mail.Service.Entity.EmailBoxAddress>(Remotion.Linq.QueryModel queryModel = {Remotion.Linq.QueryModel}) Line 67	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.Internal.QueryCompiler.CompileQuery.AnonymousMethod__0() Line 177	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.Internal.CompiledQueryCache.GetOrAddQuery<System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress>>(object cacheKey = {Microsoft.Data.Entity.Query.Internal.SqlServerCompiledQueryCacheKeyGenerator.SqlServerCompiledQueryCacheKey}, System.Func<System.Func<Microsoft.Data.Entity.Query.QueryContext, System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress>>> compiler = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 32	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.Internal.QueryCompiler.CompileQuery<System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress>>(System.Linq.Expressions.Expression query = {System.Linq.Expressions.MethodCallExpressionN}) Line 138	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.Internal.QueryCompiler.Execute<System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress>>(System.Linq.Expressions.Expression query = {System.Linq.Expressions.MethodCallExpressionN}) Line 84	C#
 	EntityFramework.Core.dll!Microsoft.Data.Entity.Query.Internal.EntityQueryProvider.Execute<System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress>>(System.Linq.Expressions.Expression expression = {System.Linq.Expressions.MethodCallExpressionN}) Line 37	C#
 	Remotion.Linq.dll!Remotion.Linq.QueryableBase<LonDatSources.Mail.Service.Entity.EmailBoxAddress>.GetEnumerator()	Unknown
 	System.Core.dll!System.Linq.Buffer<LonDatSources.Mail.Service.Entity.EmailBoxAddress>.Buffer(System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress> source)	Unknown
 	System.Core.dll!System.Linq.Enumerable.ToArray<LonDatSources.Mail.Service.Entity.EmailBoxAddress>(System.Collections.Generic.IEnumerable<LonDatSources.Mail.Service.Entity.EmailBoxAddress> source)	Unknown
```
This working fine until today morning latest (almost sure was working yesterday) :)

Thx,
Milan

**Comment by LonDat** [comment]
TEMPORARY FIX (FOR ME)...not cover all situations!:

file: \src\EntityFramework.Core\Query\ExpressionVisitors\Internal\ **NavigationRewritingExpressionVisitor.cs**
line: **226**:
```C#
            if (newLeft.Type != newRight.Type && Nullable.GetUnderlyingType(newLeft.Type) == newRight.Type)
            {
                if (newLeft is MethodCallExpression)
                {
                    var me = (MethodCallExpression) newLeft;
                    // for constant
                    if (newRight is ConstantExpression)
                    {
                        var c = (ConstantExpression) newRight;
                        newRight = Expression.Constant(Activator.CreateInstance(typeof(Nullable<>).MakeGenericType(newRight.Type), new object[] { c.Value }), newLeft.Type);
                    }
                    // parameter
                    if (newRight is ParameterExpression)
                    {
                        var p = (ParameterExpression) newRight;
                        newRight = Expression.Parameter(newLeft.Type, p.Name);
                    }
                }
            }
```

**Comment by maumar** [comment]
Introduced in my latest checkin, will look at it today. Thanks for quickly reporting this!

**Comment by maumar** [comment]
@LonDat could you provide code for DbContext? I was able to repro this for case where relationship between X and Y was required, but the YId is a nullable<int>. I want to make sure that there is not another case that I have missed.

**Comment by maumar** [comment]
@LonDat you can take a look at proposed fix - https://github.com/aspnet/EntityFramework/pull/4045

**Comment by maumar** [comment]
@LonDat alternatively, you can simply convert the non-nullable Expression to nullable. 
We do similar thing in NavigationRewritingExpressionVisitor 457:

```
                    if (innerKeySelector.Type != joinClause.OuterKeySelector.Type)
                    {
                        if (innerKeySelector.Type.IsNullableType())
                        {
                            joinClause.OuterKeySelector
                                = Expression.Convert(
                                    joinClause.OuterKeySelector,
                                    innerKeySelector.Type);
                        }
                        else
                        {
                            innerKeySelector
                                = Expression.Convert(
                                    innerKeySelector,
                                    joinClause.OuterKeySelector.Type);
                        }
                    }
```

This deals with binary expression, but still breaks in other cases, e.g. if projecting x.YRef.Id in anonymous type.

**Comment by LonDat** [comment]
Hi @maumar,
Your fix working for me from .
To my model note...I have required reference with non-nullable PK on both entities and I not using anonymous types :)

Thx for quick fix.

M.

**Comment by maumar** [comment]
Fixed in a37c44b49ef3e87c8b810e68d1a28c1a756d236a
